### PR TITLE
MultivariateIntPolynomial from PR819

### PR DESCRIPTION
**Pull request by myluszczak**
This pull request contains work done in PR 819.  However, I botched the rebase and we decided that it would be simpler to make a new pull request.  This PR adds a multivariable polynomial object with integer coefficients.

### CODE REVIEW

**Comment by isuruf** [comment]
Can you remove the styling changes? And replace tabs with 4 spaces?

**Comment by myluszczak** [comment]
@isuruf Whitespace fixes added.

File: symengine/dict.cpp [hunk 1]
```diff
@@ -190,43 +190,73 @@ int map_int_Expr_compare(const map_int_Expr &A, const map_int_Expr &B)
...
+    if (vb.size() < va.size())
+        return 1;
     return 0;
 }
 
-bool multiset_basic_eq(const multiset_basic &a, const multiset_basic &b)
-{
-    return set_eq<multiset_basic>(a, b);
-}
+//coppied from umap_eq, with derefrencing of image in map removed.
```
Comment (shivamvats) [inline symengine/dict.cpp]: * Copied

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -575,4 +575,487 @@ RCP<const UnivariatePolynomial> mul_uni_poly(RCP<const UnivariatePolynomial> a,
...
+        if (target != dict.end()) {
+            target->second += bucket.second;
+        } else {
+           dict.insert(std::pair<vec_uint, integer_class>(translate(bucket.first,v2,size),bucket.second));
+        }
+    }
+    return MultivariateIntPolynomial::from_dict(s,std::move(dict));
+}
+
+// adds together translate two vec_uints to the proper format and adds them together componentwise
```
Comment (shivamvats) [inline symengine/polynomial.cpp]: This can be reworded.

File: symengine/polynomial.h [hunk 3]
```diff
@@ -275,10 +277,67 @@ class UnivariateExprPolynomial {
...
+    std::size_t __hash__() const;
+    bool __eq__(const Basic &o) const;
+    int compare(const Basic &o) const;
+    integer_class eval(std::map<RCP<const Symbol>, integer_class, RCPSymbolCompare> &vals) const;
+    std::string toString() const;
+};
+
+//reconciles the positioning of the exponents in the vectors in the umap_uvec_mpz dict_ of the arguments
+//with the positioning of the exponents in the correspondng vectors of the output of the function.
+//f1 and f2 are vectors whose indicies are the positions in the arguments and whose values are the
```
Comment (shivamvats) [inline symengine/polynomial.h]: indices

File: symengine/polynomial.h [hunk 4]
```diff
@@ -275,10 +277,67 @@ class UnivariateExprPolynomial {
...
+//positions in the output.  set_sym s is the set of symbols of the output, and
+// s1 and s2 are the sets of the symbols of the inputs.
+unsigned int reconcile(vec_uint &v1, vec_uint &v2, set_sym &s, const set_sym &s1, const set_sym &s2);
+//same as above, but for reconcileing representation of a UnivariatePolynomial.
```
Comment (shivamvats) [inline symengine/polynomial.h]: reconciling

File: symengine/polynomial.h [hunk 5]
```diff
@@ -275,10 +277,67 @@ class UnivariateExprPolynomial {
...
+unsigned int reconcile(vec_uint &v1, unsigned int &v2, set_sym &s, const set_sym &s1, const RCP<const Symbol> s2);
+//translates vectors from one polynomial into vectors for another.
+vec_uint translate(vec_uint original, vec_uint translator, unsigned int size);
+//tramslates terms of UnivariateIntPolynomial into vectors
```
Comment (shivamvats) [inline symengine/polynomial.h]: translates

File: symengine/tests/basic/test_polynomial.cpp [hunk 6]
```diff
@@ -302,3 +304,197 @@ TEST_CASE("Univariate Polynomial expand", "[UnivariatePolynomial][expand]")
...
+    REQUIRE(add_mult_poly(*p2,*p1)->__str__() == "- 2*x**3 y**2 z + x y**2 z**3 + y z**2 + 3*x**2 + 2*x + 3");
+    REQUIRE(sub_mult_poly(*p1,*p2)->__str__() == "- 2*x**3 y**2 z + x y**2 z**3 + y z**2 + x**2 + 3");
+    REQUIRE(sub_mult_poly(*p2,*p1)->__str__() == "2*x**3 y**2 z - x y**2 z**3 - y z**2 - x**2 - 3");
+    REQUIRE(mul_mult_poly(*p1,*p2)->__str__() == "- 2*x**5 y**2 z + x**3 y**2 z**3 - 2*x**4 y**2 z + x**2 y**2 z**3 + x**2 y z**2 + 2*x**4 + x y z**2 + 3*x**3 + 4*x**2 + 3*x");
+    REQUIRE(mul_mult_poly(*p2,*p1)->__str__() == "- 2*x**5 y**2 z + x**3 y**2 z**3 - 2*x**4 y**2 z + x**2 y**2 z**3 + x**2 y z**2 + 2*x**4 + x y z**2 + 3*x**3 + 4*x**2 + 3*x");
+
+    REQUIRE(add_mult_poly(*p1,*p3)->__str__() == "- 2*x**3 y**2 z + x y**2 z**3 + y z**2 + 2*x**2 + y**2 + x + y + 3");
+}
+
+TEST_CASE("Testing addition, subtraction, multiplication of MultivariteIntPolynomials with a UnivariateIntPolynomial whose variables are not in the variable set", "[MultivariateIntPolynomial][UnivariateIntPolynomial]")
```
Comment (shivamvats) [inline symengine/tests/basic/test_polynomial.cpp]: Multivariate

File: symengine/tests/basic/test_polynomial.cpp [hunk 7]
```diff
@@ -302,3 +304,197 @@ TEST_CASE("Univariate Polynomial expand", "[UnivariatePolynomial][expand]")
...
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Symbol> y = symbol("y");
+    RCP<const Symbol> z = symbol("z");
+    RCP<const MultivariateIntPolynomial> p = MultivariateIntPolynomial::from_dict({x,y}, { {{2,1},3_z}, {{1,2},2_z}, {{2,0},3_z}, {{0,2},2_z}, {{1,0}, 3_z}, {{0,1},2_z}, {{0,2},2_z}, {{0,0},5_z}  });
+    REQUIRE(p->diff(x)->__str__() == "6*x y + 2*y**2 + 6*x + 3");
+    REQUIRE(p->diff(y)->__str__() == "3*x**2 + 4*x y + 4*y + 2");
+    REQUIRE(p->diff(z)->__str__() == "0");
+}
+
+TEST_CASE("Testing addition, subtraction, multiplication of MultivariteIntPolynomials with a UnivariateIntPolynomial whose variable are in the variable set", "[MultivariateIntPolynomial][UnivariateIntPolynomial]")
```
Comment (shivamvats) [inline symengine/tests/basic/test_polynomial.cpp]: Multivariate

File: symengine/tests/basic/test_polynomial.cpp [hunk 8]
```diff
@@ -302,3 +304,197 @@ TEST_CASE("Univariate Polynomial expand", "[UnivariatePolynomial][expand]")
...
+    REQUIRE(p3->__eq__( *add_mult_poly(*p1, *p2) ));
+
+}
+
+TEST_CASE("Testing MultivariateIntPolynomial::eval((std::map<RCP<const Symbol>, integer_class, RCPSymbolCompare> &vals)", "[MultivariateIntPolynomial]")
+{
+    RCP<const Symbol> x = symbol("x");
+    RCP<const Symbol> y = symbol("y");
+    RCP<const Symbol> z = symbol("z");
+    RCP<const MultivariateIntPolynomial> p = MultivariateIntPolynomial::from_dict( {x,y,z}, { {{2,0,0},1_z}, {{0,2,0},2_z}, {{0,0,2},3_z}, {{1,1,1},4_z}, {{1,1,0},1_z}, {{0,1,1},2_z}, {{1,0,0},1_z}, {{0,1,0},2_z}, {{0,0,1}, 3_z} , {{0,0,0},5_z} });
```
Comment (shivamvats) [inline symengine/tests/basic/test_polynomial.cpp]: Fewer characters per line help in code review. I follow 80 chars per line; you could choose your own limit.

**Comment by shivamvats** [comment]
@StoicBronco @irislq @chenchfort Could you have a look too?

**Comment by myluszczak** [comment]
@shivamvats I have fixed the spelling errors and have reduced the width of most lines to below 100 characters.  However, some lines contain string literals with more than 100 characters, and I did not want to break up the string literals.

**Comment by myluszczak** [comment]
On travis CI build 19, compilation stops at 
`[  7%] Building CXX object symengine/utilities/teuchos/CMakeFiles/teuchos.dir/Teuchos_stacktrace.cpp.o`
with error message 
`No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
The build has been terminated`

**Comment by chenchfort** [comment]
@myluszczak Just need to restart the build.

File: symengine/derivative.cpp [hunk 9]
```diff
@@ -466,6 +466,28 @@ static RCP<const Basic> diff(const CLASS &self, \
         }
     }
 
+    static RCP<const Basic> diff(const MultivariateIntPolynomial &self, const RCP<const Symbol> &x) {
```
Comment (chenchfort) [inline symengine/derivative.cpp]: Space between if and (.
Space between ) and {.

File: symengine/derivative.cpp [hunk 10]
```diff
@@ -466,6 +466,28 @@ static RCP<const Basic> diff(const CLASS &self, \
...
+        if(self.vars_.find(x) != self.vars_.end()){
```
Comment (chenchfort) [inline symengine/derivative.cpp]: Space between if and (.
Space between ) and {.

File: symengine/derivative.cpp [hunk 11]
```diff
@@ -466,6 +466,28 @@ static RCP<const Basic> diff(const CLASS &self, \
...
+            umap_uvec_mpz dict;
+            auto i = self.vars_.begin();
+            unsigned int index = 0;
+            while(!(*i)->__eq__(*x)) {
```
Comment (chenchfort) [inline symengine/derivative.cpp]: Remove extra space between ) and {.

File: symengine/derivative.cpp [hunk 12]
```diff
@@ -466,6 +466,28 @@ static RCP<const Basic> diff(const CLASS &self, \
...
+                i++;
+                index++;
+            } //find the index of the variable we are differentiating WRT.
+            for(auto bucket : self.dict_) {
```
Comment (chenchfort) [inline symengine/derivative.cpp]: Same here

File: symengine/derivative.cpp [hunk 13]
```diff
@@ -466,6 +466,28 @@ static RCP<const Basic> diff(const CLASS &self, \
...
+                if(bucket.first[index] != 0) {
+                    vec_uint v = bucket.first;
+                    v[index]--;
+                    dict.insert(std::pair<vec_uint, integer_class>(v, bucket.second * bucket.first[index]));
+                }
+            }
+        return MultivariateIntPolynomial::from_dict(self.vars_, std::move(dict));
```
Comment (chenchfort) [inline symengine/derivative.cpp]: Indent.

File: symengine/derivative.cpp [hunk 14]
```diff
@@ -466,6 +466,28 @@ static RCP<const Basic> diff(const CLASS &self, \
...
+        if(self.vars_.find(x) != self.vars_.end()){
+            umap_uvec_mpz dict;
+            auto i = self.vars_.begin();
+            unsigned int index = 0;
+            while(!(*i)->__eq__(*x)) {
+                i++;
+                index++;
+            } //find the index of the variable we are differentiating WRT.
+            for(auto bucket : self.dict_) {
+                if(bucket.first[index] != 0) {
```
Comment (chenchfort) [inline symengine/derivative.cpp]: Same here.

File: symengine/dict.cpp [hunk 15]
```diff
@@ -190,43 +190,73 @@ int map_int_Expr_compare(const map_int_Expr &A, const map_int_Expr &B)
...
+
+    for (unsigned int i = 0; i < va.size() && i < vb.size(); i++) {
+        if (vec_uint_compare()(va[i], vb[i])) {
+            return -1;
+        } else if (!vec_uint_compare()(va[i], vb[i]) && va[i] != vb[i]) {
+            return 1;
+        } else {
+            if (a.find(va[i])->second != b.find(vb[i])->second) {
+                if(a.find(va[i])->second < b.find(vb[i])->second) {
+                    return -1;
```
Comment (chenchfort) [inline symengine/dict.cpp]: Same here.

File: symengine/dict.h [hunk 16]
```diff
@@ -139,6 +143,107 @@ typedef struct
...
+};
+
+typedef std::set< RCP<const Symbol>, RCPSymbolCompare> set_sym;
+typedef std::unordered_map<RCP<const Symbol>, unsigned int, RCPSymbolHash, RCPSymbolEq>
+    umap_sym_uint;
+typedef std::unordered_map<vec_uint, integer_class, vec_uint_hash, vec_uint_eq> umap_uvec_mpz;
+
+//Takes an unordered map of type M with key type K and returns a vector of K ordered by C.
+template<class K, class M, class C>
+    std::vector<K> order_umap(const M &d){
```
Comment (chenchfort) [inline symengine/dict.h]: Space between ) and {

File: symengine/polynomial.cpp [hunk 17]
```diff
@@ -575,4 +575,516 @@ RCP<const UnivariatePolynomial> mul_uni_poly(RCP<const UnivariatePolynomial> a,
...
+}
+
+RCP<const MultivariateIntPolynomial> MultivariateIntPolynomial::from_dict(const set_sym &s,
+        umap_uvec_mpz &&d) {
+    umap_sym_uint degs;
+    auto itter = d.begin();
+    while (itter != d.end()) {
+        if (integer_class(0) == itter->second) {
+            auto toErase = itter;
+            itter++;
```
Comment (chenchfort) [inline symengine/polynomial.cpp]: iter

**Comment by irislq** [comment]
Don't see anything wrong with the code so +1

File: symengine/tests/basic/test_polynomial.cpp (hunk 6 above)
Comment (chenchfort) [inline symengine/tests/basic/test_polynomial.cpp]: ping @myluszczak

File: symengine/polynomial.h [hunk 18]
```diff
@@ -275,10 +277,82 @@ class UnivariateExprPolynomial {
...
+    std::size_t __hash__() const;
+    bool __eq__(const Basic &o) const;
+    int compare(const Basic &o) const;
+    integer_class eval(std::map<RCP<const Symbol>, integer_class, RCPSymbolCompare> &vals) const;
+    std::string toString() const;
+};
+
+//reconciles the positioning of the exponents in the vectors in the umap_uvec_mpz dict_ of the arguments
+//with the positioning of the exponents in the correspondng vectors of the output of the function.
+//f1 and f2 are vectors whose indices are the positions in the arguments and whose values are the
```
Comment (shivamvats) [inline symengine/polynomial.h]: I think you mean v1 and v2.

Comment (shivamvats) [inline symengine/polynomial.h]: I couldn't understand what it does.

File: symengine/polynomial.cpp [hunk 19]
```diff
@@ -575,4 +575,516 @@ RCP<const UnivariatePolynomial> mul_uni_poly(RCP<const UnivariatePolynomial> a,
...
+        }
+    }
+    return poscount;
+}
+
+vec_uint translate(vec_uint original, vec_uint translator, unsigned int size) {
+    vec_uint changed;
+    changed.resize(size, 0);
+    for (unsigned int i = 0; i < original.size(); i++) {
+        changed[translator[i]] = original[i];
```
Comment (shivamvats) [inline symengine/polynomial.cpp]: What does this do?

**Comment by myluszczak** [comment]
@shivamvats The `translate` function takes a vector representing the exponents in a term in a polynomial in one set of variables (say x,y) and translates that to an equivalent vector representing the exponents for a polynomial in another set of variables (say x,y,z).  This is necessary because we might want to add together polynomials with different sets of variables.

The translator vector contains the indices in the new vector corresponding to each index of the old vector, and size is the size of the new vectors.  The `reconcile` function generates the translator vectors and the variable set of the new polynomial.

**Comment by shivamvats** [comment]
Oh cool! You could improve the function description in the header. 
Other than that, +1 from my side.

**Comment by myluszczak** [comment]
@shivamvats This last commit both resolves merge conflicts that have emerged between now and your last comment and improves the comment on `translate` in polynomial.h

File: symengine/polynomial.cpp [hunk 20]
```diff
@@ -661,4 +661,516 @@ RCP<const UnivariatePolynomial> mul_uni_poly(RCP<const UnivariatePolynomial> a,
...
+    return seed;
+}
+
+bool MultivariateIntPolynomial::__eq__(const Basic &o) const {
+    return ( set_eq<set_sym>(vars_, static_cast<const MultivariateIntPolynomial &>(o).vars_)
+        && umap_uvec_mpz_eq(dict_, static_cast<const MultivariateIntPolynomial &>(o).dict_) );
+}
+
+int MultivariateIntPolynomial::compare(const Basic &o) const {
+    //copied from UnivariateIntPolynomial::compare and then modified.
```
Comment (chenchfort) [inline symengine/polynomial.cpp]: Don't need this comment.

**Comment by shivamvats** [comment]
@isuruf Do we need any more changes before merging?

File: symengine/derivative.cpp [hunk 21]
```diff
@@ -512,6 +512,28 @@ class DiffImplementation
...
+            for (auto bucket : self.dict_) {
+                if (bucket.first[index] != 0) {
+                    vec_uint v = bucket.first;
+                    v[index]--;
+                    dict.insert(std::pair<vec_uint, integer_class>(v, bucket.second * bucket.first[index]));
+                }
+            }
+            return MultivariateIntPolynomial::from_dict(self.vars_, std::move(dict));
+        } else {
+            return zero;
```
Comment (isuruf) [inline symengine/derivative.cpp]: Return a MultivatiateIntPolynomial here

File: symengine/polynomial.cpp [hunk 22]
```diff
@@ -647,4 +647,516 @@ RCP<const UnivariatePolynomial> mul_uni_poly(RCP<const UnivariatePolynomial> a,
...
+            iter++;
+        }
+    }
+
+    int whichvar = 0;
+    for (auto sym : s) {
+        degs.insert(std::pair<RCP<const Symbol>, unsigned int>(sym,0));
+        for (auto bucket : d) {
+            if (bucket.first[whichvar] > degs.find(sym)->second)
+                degs.find(sym)->second = bucket.first[whichvar];
```
Comment (isuruf) [inline symengine/polynomial.cpp]: What does this do here?

**Comment by myluszczak** [comment]
@isuruf Those lines set the degrees of each variable to the maximum exponent for that variable.  I have added comments to from_dict in the latest commit.

**Comment by myluszczak** [comment]
@isuruf I have submitted a pull request containing a multivariable polynomial object with expression coefficients (PR  #913).  PR #913 also contains all of the code contained in this PR.

**Comment by myluszczak** [comment]
Closed in favor of PR #913

**Pull request closed without merging**